### ISSUE TITLE
Quectel WWAN driver always found again even after installing it

### ISSUE THREAD

**Issue by SynMaki** [jantari/LSUClient#59]
### Version

1.5.0

### Computer Model

See Problem

### Problem

Hi jantari,

First of, thanks for releasing the 1.5.0 Version. Just implemented it to our Pilot Group!

Unfortunately, i have two other Problems (also had it on the previous Version). 
Problem 1: It seems, that some Drivers want to install themself on every check again (we tried reebots and shutdowns on these machines).

Problem 2: Some Chipset Device Software Drivers will not be installed.

Our Script we scheduling while after a login:
```
Start-Transcript -Path '....\LSUClientALL.txt' -Append
Get-Date
# Set a maximum allowed installer runtime of 20 minutes
Set-LSUClientConfiguration -MaxInstallerRuntime (New-TimeSpan -Minutes 20)
# Check Current Configuration
Get-LSUClientConfiguration

# Find updates
$updates = Get-LSUpdate
# Download them all to the local disk
$updates | Save-LSUpdate -Path "C:\ProgramData\LSUTemp" -Verbose
# Then install
$updates | Install-LSUpdate -Path "C:\ProgramData\LSUTemp" -Verbose
Stop-Transcript
```

**_Problem1:_**
**DEVICE 1** Model: 21C8S0MX00
Freitag, 4. November 2022 10:30:49

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

AUSFÜHRLICH: Processing package 'r1yfu05w - Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Installing package r1yfu05w ...

ID             : r1yfu05w
Title          : Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 0
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:01.3362134

------

Freitag, 11. November 2022 06:59:58

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

AUSFÜHRLICH: Processing package 'r1yfu05w - Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Installing package r1yfu05w ...

ID             : r1yfu05w
Title          : Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 0
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:02.1884399

--------

Montag, 14. November 2022 07:06:43

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

AUSFÜHRLICH: Processing package 'r1yfu05w - Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Installing package r1yfu05w ...

ID             : r1yfu05w
Title          : Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 0
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:01.3805968

---------------

Montag, 26. September 2022 08:12:08
AUSFÜHRLICH: Processing package 'n24au05w - Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Starting download of 'https://download.lenovo.com/pccbbs/mobiles/n24au05w.txt'
AUSFÜHRLICH: Installing package n24au05w ...

ID             : n24au05w
Title          : Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 259
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:18.1586815

--------

**Device 2** Model: Mdl20L6SBLF00

Freitag, 11. November 2022 06:47:12

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

AUSFÜHRLICH: Processing package 'n24au05w - Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Starting download of 'https://download.lenovo.com/pccbbs/mobiles/n24au05w.txt'
AUSFÜHRLICH: Installing package n24au05w ...

ID             : n24au05w
Title          : Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 259
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:16.6687355

------

Montag, 14. November 2022 07:00:56

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

AUSFÜHRLICH: Processing package 'n24au05w - Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Starting download of 'https://download.lenovo.com/pccbbs/mobiles/n24au05w.txt'
AUSFÜHRLICH: Installing package n24au05w ...

ID             : n24au05w
Title          : Realtek Audio Driver - 10 (1809 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : True
FailureReason  :
PendingAction  : REBOOT_SUGGESTED
ExitCode       : 259
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:16.5505164

------

_**Problem 2**_

Model: Mdl20SYS3C802

Montag, 14. November 2022 06:46:28

Proxy                       :
ProxyCredential             :
ProxyUseDefaultCredential   : False
MaxExternalDetectionRuntime : 00:10:00
MaxExtractRuntime           : 00:20:00
MaxInstallerRuntime         : 00:20:00

Ausnahme beim Aufrufen von "DownloadFile" mit 2 Argument(en):  "Der Remoteserver hat einen Fehler zurückgegeben: (404) 
Nicht gefunden."
In C:\Program Files\WindowsPowerShell\Modules\LSUClient\private\Save-PackageFile.ps1:46 Zeichen:9
+         $webClient.DownloadFile($SourceFile.AbsoluteLocation, $Downlo ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException
Ausnahme beim Aufrufen von "DownloadFile" mit 2 Argument(en):  "Der Remoteserver hat einen Fehler zurückgegeben: (404)
Nicht gefunden."
In C:\Program Files\WindowsPowerShell\Modules\LSUClient\private\Save-PackageFile.ps1:46 Zeichen:9
+         $webClient.DownloadFile($SourceFile.AbsoluteLocation, $Downlo ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : WebException

WARNUNG: The command or file '%PACKAGEPATH%\tbtfwdetect26_dch.exe 8' could not be found from 'C:\Windows\TEMP\djjy5s5b.crl\lpctmp1o.qkw' and was not run
AUSFÜHRLICH: Processing package 'n2uic02w - Intel Chipset Device Softwar - 10 (1909 or Later)/11 (21H2 or Later)'
AUSFÜHRLICH: Starting download of 'https://download.lenovo.com/pccbbs/mobiles/n2uic02w.txt'
AUSFÜHRLICH: Installing package n2uic02w ...

ID             : n2uic02w
Title          : Intel Chipset Device Softwar - 10 (1909 or Later)/11 (21H2 or Later)
Type           : Driver
Success        : False
FailureReason  : INSTALLER_EXITCODE
PendingAction  : NONE
ExitCode       : 13312
StandardOutput : {}
StandardError  : {}
LogOutput      : {}
Runtime        : 00:00:03.7725961

Thanks for your Time here!

**Comment by jantari** [comment]
Hi,

to troubleshoot problem 1, please run `Get-LSUpdate -Verbose -Debug` after the driver has already been installed once.
With `-Debug` this will produce a lot of output, but it will detail exactly why each driver package is considered installed/not installed so if there is something wrong with that detection we can see it.

For problem 2, that package uses `DPInst.exe` as its installer:

```
$20sy_updates | Where-Object ID -eq 'n2uic02w' | Select-Object -ExpandProperty Installer

Unattended     : True
InstallType    : cmd
SuccessCodes   : {0, 1}
InfFile        :
ExtractCommand : n2uic02w.exe /VERYSILENT /DIR=%PACKAGEPATH% /EXTRACT="YES"
Command        : %PACKAGEPATH%\DPInst.exe /sa /se /sw
```

which is an (older) official Microsoft tool, and there is some documentation around it. For example [this website](https://www.cmehren.de/2016/05/13/dpinst-treiber-installation/) explains what the return codes mean.

You are getting `13312` which, translated to hexadecimal, is `00003400`.

If the information in the above link is correct this would mean 34 drivers were added to Windows successfully, but none of them could directly be used for a connected device.

I think it could be possible that the drivers are installed but not used because they aren't really applicable to your machine. Maybe this is a similar problem to problem 1, where a package is being mis-detected. If you could run `Get-LSUpdate -Verbose -Debug` on this model as well that would be helpful to see if that's the case.

Thanks!

**Comment by jantari** [comment]
I have had some time to look into this and have a few ideas on what could be the cause of it, but have to do more testing to verify.

I also identified the same issue in some logfiles from model 20X1 deployments with a similar Quectel WWAN driver `r1kwq15w` so I should be able to reproduce it on those computers to find out more. Will keep you updated.

**Comment by jantari** [comment]
So, I had an idea how to fix this and implemented it in dbee40c but I couldn't validate the fix on any 20X3 anymore because Lenovo has updated the package in the meantime and superseded [the one that originally had the problem for me](https://download.lenovo.com/pccbbs/mobiles/r1kwq15w_2_.xml) with [a newer version](https://download.lenovo.com/pccbbs/mobiles/r1kwq19w_2_.xml) that works differently and

**Comment by jantari** [comment]
I could compare the results from the current release 1.5.0 vs develop on a 21C5 today with the exact driver package `r1yfu05w` you have too. I can now confirm the problem is fixed in develop.

1.5.0 results:

```
ID       RebootType IsApplicable IsInstalled Title
--       ---------- ------------ ----------- -----
fwnva58           5         True       False Lenovo NVMe Solid State Drive Firmware Update Utility - 10/11
r1yfu05w          3         True       False Quectel EM05-G Wireless WAN Driver - 10 (20H2 or Later)/11 (21H2 or Later)
r1yuj07w          5         True       False ThinkPad BIOS Update - 10/11
```

develop-branch results as of d595dcb:

```
ID       RebootType IsApplicable IsInstalled Title
--       ---------- ------------ ----------- -----
fwnva58           5         True       False Lenovo NVMe Solid State Drive Firmware Update Utility - 10/11
r1yuj07w          5         True       False ThinkPad BIOS Update - 10/11
```

with `r1yfu05w` now reporting as IsApplicable TRUE and IsInstalled TRUE.

**Comment by jantari** [comment]
Published Version 1.5.1 which includes this fix
